src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/TupleTypeDecoder.vb

```diff
@@ -0,0 +1,308 @@
+                    '
+                    ' The process for decoding tuples in looks at the BaseType, calls
+                    ' DecodeOrThrow, then passes the decoded type to the TupleTypeDecoder.
+                    ' However, DecodeOrThrow uses the AbstractTypeMap to construct a
+                    ' SubstitutedTypeSymbol, which eagerly converts tuple-compatible
+                    ' types to TupleTypeSymbols. Thus, by the time we get to the Decoder
+                    ' all metadata instances of System.ValueTuple will have been
+                    '  replaced with TupleTypeSymbols without names.
+                    ' 
+                    ' Rather than fixing up after-the-fact it's possible that we could
+                    ' flow up a SubstituteWith/Without tuple unification to the top level
+                    ' of the type map and change DecodeOrThrow to call into the substitution
+                    ' without unification instead.
+                    Return If(type.IsTupleType,
+                             DecodeNamedType(type.TupleUnderlyingType),
+                             DecodeNamedType(DirectCast(type, NamedTypeSymbol)))
+
+                Case SymbolKind.ArrayType
+                    Return DecodeArrayType(DirectCast(type, ArrayTypeSymbol))
+
+                Case Else
+                    Throw ExceptionUtilities.UnexpectedValue(type.TypeKind)
+            End Select
+        End Function
+
+        Private Function DecodeNamedType(type As NamedTypeSymbol) As NamedTypeSymbol
+            ' First decode the type arguments
+            Dim typeArgs = type.TypeArgumentsNoUseSiteDiagnostics
+            Dim decodedArgs = DecodeTypeArguments(typeArgs)
+
+            Dim decodedType = type
+
+            ' Now check the container
+            Dim containingType = type.ContainingType
+            Dim decodedContainingType As NamedTypeSymbol = Nothing
+
+            If containingType IsNot Nothing AndAlso containingType.IsGenericType Then
+                decodedContainingType = DecodeNamedType(containingType)
+                Debug.Assert(decodedContainingType.IsGenericType)
+            Else
+                decodedContainingType = containingType
+            End If
+
+            ' Replace the type if necessary
+            Dim containerChanged = decodedContainingType IsNot containingType
+            Dim typeArgsChanged = typeArgs <> decodedArgs
+            If typeArgsChanged OrElse containerChanged Then
+                Dim newTypeArgs = If(type.HasTypeArgumentsCustomModifiers,
+                                     decodedArgs.ZipAsArray(type.TypeArgumentsCustomModifiers, Function(t, m) New TypeWithModifiers(t, m)),
+                                     decodedArgs.SelectAsArray(Function(t) New TypeWithModifiers(t, Nothing)))
+
+                If (containerChanged) Then
+                    decodedType = decodedType.OriginalDefinition.AsMember(decodedContainingType)
```

Comment:
was supposed to be used instead of type 

---
In reply to: [77901955](https://github.com/dotnet/roslyn/pull/13669#discussion_r77901955) [](ancestors = 77901955)